pkg/addon/render.go

```diff
@@ -0,0 +1,278 @@
+/*
+Copyright 2022 The KubeVela Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package addon
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"path"
+	"strings"
+
+	cueyaml "cuelang.org/go/encoding/yaml"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+
+	common2 "github.com/oam-dev/kubevela/apis/core.oam.dev/common"
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1alpha1"
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1beta1"
+	"github.com/oam-dev/kubevela/apis/types"
+	cuemodel "github.com/oam-dev/kubevela/pkg/cue/model"
+	"github.com/oam-dev/kubevela/pkg/cue/model/value"
+	"github.com/oam-dev/kubevela/pkg/multicluster"
+	"github.com/oam-dev/kubevela/pkg/oam"
+	addonutil "github.com/oam-dev/kubevela/pkg/utils/addon"
+)
+
+type addonCueTemplateRender struct {
+	addon     *InstallPackage
+	inputArgs map[string]interface{}
+}
+
+// This func can be used for addon render, supporting render app template and component.
+// Please notice the result will be stored in object parameter, so object must be a pointer type
+func (a addonCueTemplateRender) toObject(cueTemplate string, object interface{}) error {
```

Comment:
```suggestion
func (a addonCueTemplateRender) render(cueTemplate string) (*Application,error) {
```

This function can only output Application cause you have specific logic inside.

Reply:
This func can be called by both render app and appCompnent, that's determined by in put object.

Reply:
if so ,you should make the cuepath `output` as an input, it can become much more common function

Reply:
Why we need change this the path? The appComponent and app are all in `output` block.

Reply:
how would you deal with the case for "outputs"?

Reply:
Since the input parameter is cueTemplate file will contain one output and several outputs, and the outputs is unstructed object, I plan to add a input parameter value as `aliaOutputs []unstructed.unstructure` which fetch all outpus in this cueTemplate